File: packages/core/src/decay-worker.ts
import type { Redis } from 'ioredis'
import { KeyBuilder } from './key-builder.js'
import type { PenaltyThresholds } from './types.js'
import { DEFAULT_THRESHOLDS } from './penalty-state-machine.js'

// ---------------------------------------------------------------------------
// Lua script — identical to engine/internal/store/decay.go scoreDecayLua
// ---------------------------------------------------------------------------
const LUA_DECAY = `
local score_key    = KEYS[1]
local state_key    = KEYS[2]
local decay_ts_key = KEYS[3]
local now          = tonumber(ARGV[1])
local half_life_ms = tonumber(ARGV[2])
local warn_t       = tonumber(ARGV[3])
local slow_t       = tonumber(ARGV[4])
local block_t      = tonumber(ARGV[5])
local blacklist_t  = tonumber(ARGV[6])

local raw = redis.call('GET', score_key)
if not raw then return {0, 'CLEAN', 0} end
local score = tonumber(raw)
if score <= 0 then return {0, 'CLEAN', 0} end

local last_decay = tonumber(redis.call('GET', decay_ts_key) or tostring(now))
local elapsed = now - last_decay
if elapsed <= 0 then
  return {score, redis.call('GET', state_key) or 'CLEAN', 0}
end

local factor    = math.exp(-0.693147 * elapsed / half_life_ms)
local new_score = math.floor(score * factor)
if new_score < 0 then new_score = 0 end
local decrement = score - new_score

redis.call('SET', decay_ts_key, now)

if decrement <= 0 then
  return {score, redis.call('GET', state_key) or 'CLEAN', 0}
end

redis.call('SET', score_key, new_score)

local state
if     new_score >= blacklist_t then state = 'BLACKLIST'
elseif new_score >= block_t     then state = 'BLOCK'
elseif new_score >= slow_t      then state = 'SLOW'
elseif new_score >= warn_t      then state = 'WARN'
else                                  state = 'CLEAN'
end

redis.call('SET', state_key, state)
return {new_score, state, decrement}
`

export interface DecayResult {
  ip: string
  newScore: number
  newState: string
  decrement: number
}

export interface DecayWorkerOptions {
  /** Half-life for exponential score decay. Default: 10 minutes. */
  halfLifeMs?: number
  /** How often to run a decay cycle. Default: 60 000 ms (1 minute). */
  intervalMs?: number
  thresholds?: PenaltyThresholds
  onDecay?: (results: DecayResult[]) => void
  onError?: (err: Error) => void
}

/**
 * DecayWorker periodically scans penalty score keys in Redis and applies
 * exponential half-life decay — allowing IPs to "cool down" over time
 * without requiring the Go engine.
 *
 * Usage (standalone / no engine):
 *   const worker = new DecayWorker(redis, { tenant: 'my-app' })
 *   worker.start()
 *   // on shutdown:
 *   worker.stop()
 */
export class DecayWorker {
  private timer?: ReturnType<typeof setInterval>
  private readonly keys: KeyBuilder
  private readonly halfLifeMs: number
  private readonly intervalMs: number
  private readonly thresholds: PenaltyThresholds

  constructor(
    private readonly redis: Redis,
    private readonly tenant: string,
    private readonly opts: DecayWorkerOptions = {},
  ) {
    this.keys = new KeyBuilder(tenant)
    this.halfLifeMs = opts.halfLifeMs ?? 10 * 60 * 1000
    this.intervalMs = opts.intervalMs ?? 60_000
    this.thresholds = opts.thresholds ?? DEFAULT_THRESHOLDS
  }

  /** Start periodic decay. Fires immediately, then every intervalMs. */
  start(): void {
    if (this.timer !== undefined) return
    const tick = (): void => {
      this.runCycle().catch(err => this.opts.onError?.(err as Error))
    }
    tick()
    this.timer = setInterval(tick, this.intervalMs)
    if (typeof this.timer.unref === 'function') this.timer.unref()
  }

  stop(): void {
    if (this.timer !== undefined) {
      clearInterval(this.timer)
      this.timer = undefined
    }
  }

  /** Run a single decay cycle — callable directly in tests. */
  async runCycle(): Promise<DecayResult[]> {
    const ips = await this.scanIPs()
    if (ips.length === 0) return []

    const results = await Promise.all(ips.map(ip => this.decayOne(ip)))
    const changed = results.filter((r): r is DecayResult => r !== null && r.decrement > 0)

    if (changed.length > 0) this.opts.onDecay?.(changed)
    return changed
  }

  // -------------------------------------------------------------------------

  private async scanIPs(): Promise<string[]> {
    const pattern = this.keys.penaltyScorePattern('ip')
    const prefix  = this.keys.penaltyScore('ip', '')
    const ips: string[] = []

    let cursor = '0'
    do {
      const [next, keys] = await this.redis.scan(cursor, 'MATCH', pattern, 'COUNT', 100)
      cursor = next
      for (const key of keys) {
        const ip = key.slice(prefix.length)
        if (ip) ips.push(ip)
      }
    } while (cursor !== '0')

    return ips
  }

  private async decayOne(ip: string): Promise<DecayResult | null> {
    const t = this.thresholds
    const res = await this.redis.eval(
      LUA_DECAY,
      3,
      this.keys.penaltyScore('ip', ip),
      this.keys.penaltyState('ip', ip),
      this.keys.penaltyDecayTs('ip', ip),
      Date.now(),
      this.halfLifeMs,
      t.warn,
      t.slow,
      t.block,
      t.blacklist,
    ) as [number, string, number]

    if (!Array.isArray(res) || res.length < 3) return null
    return {
      ip,
      newScore:  Number(res[0]),
      newState:  String(res[1]),
      decrement: Number(res[2]),
    }
  }
}

File: packages/fastapi/autoblock/decay_worker.py
"""DecayWorker — standalone exponential score decay (no engine required)."""
from __future__ import annotations

import asyncio
import logging
from dataclasses import dataclass
from typing import Callable

import redis.asyncio as aioredis

from .key_builder import KeyBuilder
from .types import PenaltyThresholds

log = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# Lua script — identical to engine/internal/store/decay.go scoreDecayLua
# ---------------------------------------------------------------------------
_LUA_DECAY = """
local score_key    = KEYS[1]
local state_key    = KEYS[2]
local decay_ts_key = KEYS[3]
local now          = tonumber(ARGV[1])
local half_life_ms = tonumber(ARGV[2])
local warn_t       = tonumber(ARGV[3])
local slow_t       = tonumber(ARGV[4])
local block_t      = tonumber(ARGV[5])
local blacklist_t  = tonumber(ARGV[6])

local raw = redis.call('GET', score_key)
if not raw then return {0, 'CLEAN', 0} end
local score = tonumber(raw)
if score <= 0 then return {0, 'CLEAN', 0} end

local last_decay = tonumber(redis.call('GET', decay_ts_key) or tostring(now))
local elapsed = now - last_decay
if elapsed <= 0 then
  return {score, redis.call('GET', state_key) or 'CLEAN', 0}
end

local factor    = math.exp(-0.693147 * elapsed / half_life_ms)
local new_score = math.floor(score * factor)
if new_score < 0 then new_score = 0 end
local decrement = score - new_score

redis.call('SET', decay_ts_key, now)

if decrement <= 0 then
  return {score, redis.call('GET', state_key) or 'CLEAN', 0}
end

redis.call('SET', score_key, new_score)

local state
if     new_score >= blacklist_t then state = 'BLACKLIST'
elseif new_score >= block_t     then state = 'BLOCK'
elseif new_score >= slow_t      then state = 'SLOW'
elseif new_score >= warn_t      then state = 'WARN'
else                                  state = 'CLEAN'
end

redis.call('SET', state_key, state)
return {new_score, state, decrement}
"""

_DEFAULT_HALF_LIFE_MS = 10 * 60 * 1000  # 10 minutes
_DEFAULT_INTERVAL_S   = 60               # 1 minute


@dataclass
class DecayResult:
    ip: str
    new_score: int
    new_state: str
    decrement: int


class DecayWorker:
    """
    Periodically scans penalty score keys in Redis and applies exponential
    half-life decay so IPs "cool down" over time without needing the Go engine.

    Usage::

        worker = DecayWorker(redis, tenant="my-app")
        await worker.start()
        # on shutdown:
        await worker.stop()

    Or as an async context manager::

        async with DecayWorker(redis, tenant="my-app"):
            await run_app()
    """

    def __init__(
        self,
        redis: aioredis.Redis,
        tenant: str,
        *,
        half_life_ms: int = _DEFAULT_HALF_LIFE_MS,
        interval_seconds: int = _DEFAULT_INTERVAL_S,
        thresholds: PenaltyThresholds | None = None,
        on_decay: Callable[[list[DecayResult]], None] | None = None,
        on_error: Callable[[Exception], None] | None = None,
    ) -> None:
        self._redis = redis
        self._keys = KeyBuilder(tenant)
        self._half_life_ms = half_life_ms
        self._interval = interval_seconds
        self._thresholds = thresholds or PenaltyThresholds()
        self._on_decay = on_decay
        self._on_error = on_error
        self._script = redis.register_script(_LUA_DECAY)
        self._task: asyncio.Task | None = None

    # ------------------------------------------------------------------
    # Lifecycle
    # ------------------------------------------------------------------

    async def start(self) -> None:
        if self._task is not None:
            return
        await self.run_cycle()  # immediate first decay
        self._task = asyncio.create_task(self._loop(), name="autoblock-decay-worker")

    async def stop(self) -> None:
        if self._task is not None:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
            self._task = None

    async def __aenter__(self) -> "DecayWorker":
        await self.start()
        return self

    async def __aexit__(self, *_) -> None:
        await self.stop()

    # ------------------------------------------------------------------
    # Core cycle
    # ------------------------------------------------------------------

    async def run_cycle(self) -> list[DecayResult]:
        """Run a single decay pass over all tracked IPs."""
        ips = await self._scan_ips()
        if not ips:
            return []

        tasks = [self._decay_one(ip) for ip in ips]
        results_raw = await asyncio.gather(*tasks, return_exceptions=True)
        results = [r for r in results_raw if isinstance(r, DecayResult) and r.decrement > 0]

        if results and self._on_decay:
            self._on_decay(results)
        return results

    # ------------------------------------------------------------------
    # Background loop
    # ------------------------------------------------------------------

    async def _loop(self) -> None:
        while True:
            await asyncio.sleep(self._interval)
            try:
                await self.run_cycle()
            except asyncio.CancelledError:
                raise
            except Exception as exc:
                log.warning("autoblock DecayWorker error: %s", exc)
                if self._on_error:
                    self._on_error(exc)

    # ------------------------------------------------------------------
    # Helpers
    # ------------------------------------------------------------------

    async def _scan_ips(self) -> list[str]:
        pattern = self._keys.penalty_score_pattern("ip")
        prefix  = self._keys.penalty_score("ip", "")
        ips: list[str] = []
        cursor: int = 0
        while True:
            cursor, keys = await self._redis.scan(cursor, match=pattern, count=100)
            for key in keys:
                ip = key[len(prefix):]
                if ip:
                    ips.append(ip)
            if cursor == 0:
                break
        return ips

    async def _decay_one(self, ip: str) -> DecayResult:
        t = self._thresholds
        import time
        now_ms = int(time.time() * 1000)
        res = await self._script(
            keys=[
                self._keys.penalty_score("ip", ip),
                self._keys.penalty_state("ip", ip),
                self._keys.penalty_decay_ts("ip", ip),
            ],
            args=[now_ms, self._half_life_ms, t.warn, t.slow, t.block, t.blacklist],
        )
        new_score, new_state, decrement = int(res[0]), str(res[1]), int(res[2])
        return DecayResult(ip=ip, new_score=new_score, new_state=new_state, decrement=decrement)
